### PR TITLE
Encrypt passwords + alembic!

### PR DESCRIPTION
**Pull request by mistercrunch**
Local version of
https://github.com/airbnb/airflow/pull/267

### CODE REVIEW

File: airflow/migrations/versions/e3a246e0dc1_current_schema.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            sa.Column('end_date', sa.DateTime(), nullable=True),
+            sa.Column('user_id', sa.Integer(), nullable=True),
+            sa.Column('known_event_type_id', sa.Integer(), nullable=True),
+            sa.Column('description', sa.Text(), nullable=True),
+            sa.ForeignKeyConstraint(['known_event_type_id'],
+                                    ['known_event_type.id'], ),
+            sa.ForeignKeyConstraint(['user_id'], ['user.id'], ),
+            sa.PrimaryKeyConstraint('id')
+        )
+
```
Comment (jlowin) [inline airflow/migrations/versions/e3a246e0dc1_current_schema.py:227]: Not 100% sure if it's needed -- but just a reminder about the new xcom table.

**Comment by mistercrunch** [comment]
Good catch @jlowin . Somehow I had forgotten to `git add` the latest migration file.

**Comment by neovintage** [comment]
I can merge all of this into my original or we can just merge this one. :-D

**Comment by jlowin** [comment]
@mistercrunch at least for me, this doesn't work. The first migration file `e3a246e0dc1_current_schema.py` creates the `is_encrypted` column in the `connections` table, and then `1507a7289a2f_create_is_encrypted.py` tries to add it again, resulting in:

```
ProgrammingError: (psycopg2.ProgrammingError) column "is_encrypted" of relation "connection" already exists
 [SQL: 'ALTER TABLE connection ADD COLUMN is_encrypted BOOLEAN']
```

Can't run `airflow resetdb` or `airflow initdb`.

**Comment by jlowin** [comment]
In addition, I think `airflow resetdb` needs to drop (or reset) the `alembic_version` table. Otherwise the upgrade step never runs after dropping all the airflow tables because it checks `alembic_version` and sees that all is well.

edit: looks like maybe alembic should take care of this, but perhaps it isn't happening because of the other errors?

**Comment by mistercrunch** [comment]
This should help:
https://github.com/airbnb/airflow/commit/d7554d8cc8a7c2d9227b1b76a32867a4673a82b9

**Comment by jlowin** [comment]
Unfortunately I don't think this works -- it needs an additional guard:

``` python

    if 'connection' in inspector.get_table_names(): # <<< This is the guard
        col_names = [col['name'] for col in inspector.get_columns('connection')]

        if 'is_encrypted' not in col_names:
            op.add_column(
                'connection',
                sa.Column('is_encrypted', sa.Boolean, unique=False, default=False))

            conn = op.get_bind()
            conn.execute(
                connectionhelper.update().values(is_encrypted=False)
            )
```

Alembic seems to run all the migrations before actually doing any work, so someone starting fresh (or running `resetdb`) won't have a `connection` table yet. In that case, checking the columns of `connection` raises an error.

**Comment by jlowin** [comment]
the import_error table migration is failing on my machine, for the same reason -- it gets "created" (but not actually built in the db) in the first migration version, and then the third migration version checks if it exists in the db (which it doesn't) and creates it again, resulting in two creation statements. 

I suspect that airflow can't be installed in its current form, since the `initdb` will fail.

**Comment by jlowin** [comment]
In general, I think there shouldn't be any guards in Alembic versions; the whole idea is that you can run them sequentially. This is only an issue because we need to  seamlessly migrate users from pre-alembic Airflow and that means skipping certain steps (for tables they already have).

So I think the proper way to do this is:
1. keep the first migration version largely as is, but remove the `is_encrypted` field from `connection` table, and add the `xcom` table. Then this first migration file is basically "this is what a pre-alembic Airflow user looks like" and can be full of guards against existing tables.
2. create a migration version file for adding "is_encrypted" that has no guards against it already existing. (or maybe we need one guard for users who have installed Airflow since yesterday)
3. create any future migration versions without guards...

This workflow should support existing users AND new users (including `resetdb`) without complication introduced by worrying about guards. I'll submit a PR for it if you like.

**Pull request merged**